include/oneapi/dpl/pstl/hetero/dpcpp/sycl_traits.h

```diff
@@ -0,0 +1,49 @@
+// -*- C++ -*-
+//===----------------------------------------------------------------------===//
+//
+// Copyright (C) Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// This file incorporates work covered by the following copyright and permission
+// notice:
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+//
+//===----------------------------------------------------------------------===//
+
+// This file contains some specialization SYCL traits for some oneDPL types
+//
+// Include this header before a kernel submit SYCL code
+
+#ifndef _ONEDPL_SYCL_TRAITS_H
+#define _ONEDPL_SYCL_TRAITS_H
+
+#define _ONEDPL_DEVICE_COPYABLE(TYPE) \
+template<typename... Ts > \
+struct sycl::is_device_copyable<TYPE<Ts...>>: ::std::conjunction<sycl::is_device_copyable<Ts>...> {};
+
+using namespace oneapi::dpl::__internal;
+
+_ONEDPL_DEVICE_COPYABLE(__not_pred)
+_ONEDPL_DEVICE_COPYABLE(__reorder_pred)
+_ONEDPL_DEVICE_COPYABLE(__equal_value_by_pred)
+_ONEDPL_DEVICE_COPYABLE(__equal_value)
+_ONEDPL_DEVICE_COPYABLE(__not_equal_value)
+_ONEDPL_DEVICE_COPYABLE(__transform_functor)
+_ONEDPL_DEVICE_COPYABLE(__transform_if_unary_functor)
+_ONEDPL_DEVICE_COPYABLE(__transform_if_binary_functor)
+_ONEDPL_DEVICE_COPYABLE(__replace_functor)
+_ONEDPL_DEVICE_COPYABLE(__replace_copy_functor)
+_ONEDPL_DEVICE_COPYABLE(zip_forward_iterator)
```

Comment:
I think we need to add to this list. If I am correct, effectively any functor with template types used for member variables which could be or could contain a type like `std::tuple` (which includes the other functors on this list). std::tuple is device copyable but not trivially copyable, so such a specialization would be necessary.

Here is a list of functors which seem like they may need to be added.  It may not be exhaustive.

include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h:
 `fill_functor`
 `generate_functor`
`__brick_fill`
`__brick_fill_n`
`__search_n_unary_predicate`
`__is_heap_check`

include/oneapi/dpl/pstl/hetero/utils_hetero.h:
`equal_predicate`
`adjacent_find_fn`
`__create_mask_unique_copy`

include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
`__early_exit_find_or`

include/oneapi/dpl/pstl/hetero/dpcpp/unseq_backend_sycl.h:
`walk_n`
`walk_adjacent_difference`
`transform_reduce`
`reduce_over_group`
`single_match_pred_by_idx`
`multiple_match_pred`
`n_elem_match_pred`
`first_match_pred`
`__create_mask`
`__copy_by_mask`
`__partition_by_mask`
 `__global_scan_functor`
`__scan`
`__brick_includes`
`__brick_set_op`
`__brick_reduce_idx`

include/oneapi/dpl/internal/binary_search_impl.h:
 `custom_brick` (binary_search), 

include/oneapi/dpl/internal/function.h:
 `replace_if_fun`, 
 `scan_by_key_fun` 
 `segmented_scan_fun`
 `scatter_and_accumulate_fun`
 `transform_if_stencil_fun`

include/oneapi/dpl/pstl/memory_impl.h:
 `__op_uninitialized_fill`

Reply:
@danhoeflinger,
Would you prefer to add the rest functors written above step by step in the next PRs or into this one the all mentioned functors wholly?

Reply:
I don't have a strong opinion if they go in here or in a separate PR. There is nothing we are changing in this PR which makes the status of the other functors I list worse than it is now. However I believe these functors fall into a similar category with the ones already in this PR.

Do you agree that these seem to be required?  
I believe the requirement for inclusion in the list is when a type may contain a member function of another type which is device copyable but not trivially copyable (like `std::tuple`, or any other type in this list).

Reply:
done